Add support for selectively updating manifests

vchernin opened the pull request:
Fixes https://github.com/flathub/flatpak-external-data-checker/issues/249

I opted to add an optional property to mark sources as "important", `is-important: true`. If "important" sources are present in external data, only update the manifest if at least 1 important source received an update. This basically lets apps choose when they want PRs to open, e.g. many apps may only want a PR when at least their primary upstream app source is updated. 

Manifests without important sources will not be affected by this change. f-e-d-c will continue making manifest updates as before for all modules. In other words, this does not change any existing behaviour.

There are some (I think comprehensive) tests for this feature in `test_checker.py`.

I also tested with an actual GitHub repo (forked EasyEffects which happens to have convenient action to use):

> [When is-critical-source is set for EasyEffects itself, but only libsigc++ has an update available](https://github.com/vchernin/com.github.wwmm.easyeffects/runs/5294315473?check_suite_focus=true)

Doesn't make a PR/manifest update as shown by the logs as expected

> [When is-critical-source is set for both for EasyEffects and libsigc++, and libsigc++ has an update available](https://github.com/vchernin/com.github.wwmm.easyeffects/runs/5294368744?check_suite_focus=true)

Makes [PR as expected](https://github.com/vchernin/com.github.wwmm.easyeffects/pull/5)

vchernin committed:
Add support for allowing selective manifest updates
If `--require-important-update` is passed to CLI, and a `is-important: true` and/or `is-main-source: true` source is present in external data, only update the manifest if such "important" external data received an update.
Manifests not meeting both conditions will not be affected by this change. f-e-d-c would continue making updates as before for all modules.

Use `--require-important-update` CLI argument to gate this feature. This ensures e.g. copy-pasted modules alone cannot accidentally activate this feature.

Require "important" sources to be set explictly, since there is no clear way to automatically detect which sources are "important".
Even if there was a way of detecting "important" sources, the possibility that f-e-d-c detects wrongly could be confusing to debug.

Don't reuse the appdata updater's guess that the last source is the "main" source, (done by appdata in case `is-main-source: true` is not found).
It could be confusing if the feature was enabled and f-e-d-c arbitrarily started only updating the manifest if the last source was updated.

In the future, this change could be extended with an "always-update-for-cves" setting in addition to the require-important-update setting.
Such a setting could update the manifest if a source with CVEs received an update, regardless if the updated sources are set as is-important/is-main-source.

vchernin commented on tests/test_checker.py:
```diff
@@ -135,7 +136,10 @@ async def _test_update(
-            checker._checkers = [UpdateEverythingChecker]
+
+            if not critical_source_test:
+                checker._checkers = [UpdateEverythingChecker]
```
There might be a cleaner way of handling this. The problem is if this line runs with the new tests, some failed assertions will happen. The new tests are still valid though as-is.

 I assume it's because some code here is missing handling for `type: git` sources?

vchernin replied:
Fixed by moving to a new file

gasinvein commented on src/lib/externaldata.py:
```diff
@@ -45,6 +45,7 @@
     "properties": {
         "type": {"type": "string"},
         "is-main-source": {"type": "boolean"},
+        "is-critical-source": {"type": "boolean"},
```
I'm not a fan of using the term "critical", IMO it's too strong. Maybe "important" would fit better?

vchernin replied:
I have no real opinion on this, indeed important could be used instead.

I wonder though, is there a better parameter name than `is-important-source`? Is it possible to somehow communicate the changed manifest editing/PR behaviour as part of the option? (Since at its core that's what this behaviour change is about).

e.g. `require-update-for-manifest-edits` or something? (But that one's a bit long).

gasinvein replied:
I think `is-important` would do.

gasinvein commented on src/manifest.py:
```diff
@@ -406,6 +407,29 @@ def _update_manifest(self, path, datas, changes):
             log.info("Updating %s", path)
             self._dump_manifest(path)
 
+    def has_version_changed(self, selected_data) -> bool:
```
This would be better be a method (or even a property) on `ExternalBase` objects.

gasinvein commented on src/manifest.py:
```diff
@@ -476,15 +480,42 @@ def update_manifests(self) -> t.List[str]:
+        # if is-critical-source is found, only update the manifest if at least one source with is-critical-source received an update.
+
+        # assume we need to update everything by default
+        critical_version_changed = True
```
Setting `critical_version_changed` to `True` when there are no critical sources is a bit misleading. Maybe split it to two variables, e.g. `has_critical_source` and `critical_version_changed`? Or make it tri-state, with `None` value indication no critical sources.

gasinvein commented on src/manifest.py:
```diff
@@ -476,15 +480,42 @@ def update_manifests(self) -> t.List[str]:
+        # assume we need to update everything by default
+        critical_version_changed = True
+
+        selected_data = None
```
`selected_data` isn't used outside the loop, so it's not needed.

gasinvein reviewed (changes requested):
Thanks implementing this.

On the approach in general: I believe this feature should be gated by some cli or `flathub.json` option, otherwise there is no way to update unimportant sources manually. And if gating it, probably the main source should be considered "critical" by default?

gasinvein commented on src/lib/externaldata.py:
```diff
@@ -185,6 +186,29 @@ def set_new_version(self, new_version: ExternalState, is_update=None):
                 self.state = self.State.BROKEN
             self.new_version = new_version
 
+    def has_version_changed(self) -> bool:
```
This can be a property.

gasinvein commented on src/lib/externaldata.py:
```diff
@@ -185,6 +186,29 @@ def set_new_version(self, new_version: ExternalState, is_update=None):
+            and update.version is not None
+            and (
+                (
+                    isinstance(update, ExternalFile)
```
Ideally, instead of checking type of something on `self`, `has_version_changed` method can be different on `ExternalData` and `ExternalGitRepo` classes.

gasinvein reviewed (commented):
> I'm curious do you think `test_checker.py` is reasonable as is, or should the new tests be split off into a new file?

Probably so, at least split it off to another test class.

Ideally, this test shouldn't hit network, i.e. use some dummy checker that always sets some pre-defined "new version".

vchernin commented on src/lib/checkers.py:
```diff
@@ -35,7 +36,11 @@
-    SUPPORTED_DATA_CLASSES: t.List[t.Type[ExternalBase]] = [ExternalData]
+    SUPPORTED_DATA_CLASSES: t.List[t.Type[ExternalBase]] = [
+        ExternalData,
+        ExternalGitRepo,
```
Not sure if this is correct, it was necessary to pass this assertion:

https://github.com/flathub/flatpak-external-data-checker/blob/4de54724f00e001df0986a3d01e5575abff1384a/src/lib/checkers.py#L56-L58

I assume this is since my tests happened to be git sources. If it's preferable I can simply move them to html sources or something, it shouldn't matter match.

Though I was curious, why shouldn't this assertion allow a git source? As far as I can tell, the way my git source tests work using this code makes sense (i.e. this code is operating sanely with git sources despite the original code disallowing them).

gasinvein replied:
It is incorrect. This `SUPPORTED_DATA_CLASSES` is the default, you shouldn't change it. Instead, you're supposed to redefine it on child classes, like this
```python
class MyCustomChecker(Checker):
    SUPPORTED_DATA_CLASSES = [ExternalData, ExternalGitRepo]

vchernin replied:
Makes sense, fixed

gasinvein commented on src/manifest.py:
```diff
@@ -471,20 +459,48 @@ def _update_appdata(self):
+                if data.checker_data.get(IMPORTANT_SRC_PROP) or data.checker_data.get(
+                    MAIN_SRC_PROP
+                ):
+                    selected_data = data
```
Assigning `selected_data` variable is redundant, you can just use `data` inside the loop.

vchernin replied:
thanks, fixed

gasinvein commented on tests/test_importantsource.py:
```diff
@@ -0,0 +1,606 @@
+        )
+
+
+class UpdateEverythingChecker(DummyChecker):
```
Ideally, the tests for important source should be merged into `test_checker.py` and reusing the common code instead of just copying it over.

vchernin replied:
Hm, I had previously split off the tests into a new file due to https://github.com/flathub/flatpak-external-data-checker/pull/276#pullrequestreview-893415396. 

Would you still prefer re-combining the files (and I assume also the classes)?

The tests combined are about a 1000 lines long. Also, the `DummyChecker` in `test_importantsource.py` subclasses `GitChecker` since the tests I wrote happened to use git sources. 

The tests could be switched to `extra-data` sources. Or, `test_checker.py`'s `DummyChecker` could be switched to subclass `GitChecker` (but I'm unsure if it's 100% correct for the existing `extra-data` tests in `test_checker.py`).

gasinvein replied:
>Would you still prefer re-combining the files (and I assume also the classes)?

My initial suggestion was to move the important source tests into separate class, not entire separate file. I still think it would be better, but no strong opinion on that.

vchernin replied:
> I still think it would be better, but no strong opinion on that.

I've combined the tests files, and split off `_test_update` so it can be reused by both sets of tests. There is now only a few lines of copy paste (the git checker subclassing). I needed to slightly adjust the exisiting `_test_update` calls, but they look correct.

vchernin committed:
tests: add tests for require-important-update feature
Turns _test_checker into a reusable function for both the original extra-data and the new git checker based tests.

gasinvein commented on tests/test_checker.py:
```diff
@@ -52,6 +54,66 @@
 NUM_NEW_VERSIONS = 2
 
 
+async def _test_update(
```
This is clearly a method, not a standalone function. No point to separate it from the test class. To make it reusable, attach it to a common ancestor of `TestExternalDataChecker` and `TestImportantGitExternalDataChecker`, somehow like this:
```python
class _TestWithInlineManifest(unittest.IsolatedAsyncioTestCase):
    async def _test_update(self, ...):
        ...

class TestExternalDataChecker(_TestWithInlineManifest):
    ...

class TestImportantGitExternalDataChecker(_TestWithInlineManifest):
    ...
```
Then you can call `self._test_update()` on both.

vchernin replied:
thanks, fixed

gasinvein commented on tests/test_checker.py:
```diff
@@ -52,6 +54,67 @@
+
+            self.assertEqual(new_contents, expected_new_contents)
+            if not test_important_update:
+                self.assertEqual(updates, expected_updates)
```
Making this conditional seems like a bad idea. We should verify that exactly the updates we expected were applied.

gasinvein commented on tests/test_checker.py:
```diff
@@ -52,6 +54,67 @@
+            if test_important_update:
+                checker._checkers = [GitUpdateEverythingChecker]
+            else:
+                checker._checkers = [UpdateEverythingChecker]
```
Instead of selecting a checker class based on the method argument, you can just declare the checker on the test class, so you can pass it to `checker._checkers` like this:
```suggestion
            checker._checkers = [self._DUMMY_CHECKER_CLS]
```

gasinvein commented on tests/test_checker.py:
```diff
@@ -52,6 +54,67 @@
+                self.assertEqual(releases[0].getAttribute("date"), "2019-08-28")
+            elif (
+                not test_important_update
+            ):  # releases assertions are not needed when testing important updates
```
Why? Checking which source got into appstream release entry seems important.

gasinvein commented on tests/test_checker.py:
```diff
@@ -198,10 +213,10 @@ async def test_update_json(self):
+            filename=filename,
+            contents=contents,
+            expected_new_contents=expected_new_contents,
+            expected_updates=["Update UnityHubSetup.AppImage to 1.2.3.4"],
```
Please either split unrelated changes to a separate commit or don't include them at all into this PR.
The same applies to other args->kwargs changes below.

gasinvein reviewed (changes requested):
Another round of the tests review.

Probably I can address these myself and force-push to your branch, if you don't mind.

wjt commented on src/main.py:
```diff
@@ -388,6 +388,13 @@ def parse_cli_args(cli_args=None):
+    parser.add_argument(
+        "--require-important-update",
+        help="Require an update to at least one source with is-important or is-main-source to save changes to the manifest."
+        "If no instances of is-important or is-main-source are found, assume normal behaviour and always save changes to the manifest."
```
```suggestion
        help="Require an update to at least one source with is-important or is-main-source to save changes to the manifest. "
        "If no instances of is-important or is-main-source are found, assume normal behaviour and always save changes to the manifest. "
```

wjt commented on src/manifest.py:
```diff
@@ -471,20 +459,47 @@ def _update_appdata(self):
+                        break
+
+        if found_important_update or found_important_update is None:
+
```
```suggestion
```

wjt reviewed (commented):
Great concept! and I think this design works and avoids having too much noise in the manifest.

gasinvein reviewed (changes requested):
CLI options passed to `ManifestChecker` are now wrapped in a single `CheckerOptions` object, to avoid inflating the `__init__` method. Please rebase and update options propagation accordingly.

gasinvein reviewed (approved):
I still have a bunch of nitpicks, but it feels like this PR was pending long enough already, so let's merge it.
